persistent/mapping.py

```diff
@@ -55,6 +57,17 @@ class PersistentMapping(IterableUserDict, persistent.Persistent):
     __super_pop = IterableUserDict.pop
     __super_popitem = IterableUserDict.popitem
 
+    if sys.version_info[:3] < (3, 7, 4):
+        # See PersistentList.
+        # See https://github.com/python/cpython/commit/3645d29a1dc2102fdb0f5f0c0129ff2295bcd768
+        def __copy__(self):
+            inst = self.__class__.__new__(self.__class__)
+            inst.__dict__.update(self.__dict__)
+            # Create a copy and avoid triggering descriptors
+            data = self.__dict__['data'] if 'data' in self.__dict__ else self.__dict__['_containers']
+            inst.__dict__["data"] = data.copy()
```

Comment:
I'm not sure why the `inst.update` call isn't sufficient above. This bit seems to make *two* copies of the underlying data, one under `data` and one under `_containers`.   Actually, where does `_containers` come from?

```python
Python 2.7.17 (default, Feb 11 2020, 17:51:09) 
[GCC 7.4.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from UserDict import UserDict
>>> class Foo(UserDict): pass
... 
>>> foo = Foo()
>>> foo.__dict__
{'data': {}}
```

```python
$ python3.6
Python 3.6.10 (default, Feb 11 2020, 17:35:51) 
[GCC 7.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from collections import UserDict
>>> 
>>> class Foo(UserDict): pass
... 
>>> foo = Foo()
>>> foo.__dict__
{'data': {}}
>>> foo['bar'] = 0
>>> foo.__dict__
{'data': {'bar': 0}}
```

```python
Python 3.7.6 (default, Feb 11 2020, 17:30:33) 
[GCC 7.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from collections import UserDict
>>> class Foo(UserDict): pass
... 
>>> foo = Foo()
>>> foo.__dict__
{'data': {}}
>>> foo['bar'] = 0
>>> foo.__dict__
{'data': {'bar': 0}}
```

Reply:
> I'm not sure why the inst.update call isn't sufficient above.

Because `update` makes a shallow copy, and the second object still shares `data` with the first, meaning they both mutate when you change only one. This is basically a backport of the fix for https://bugs.python.org/issue27141. (It's less of a big deal with persistent objects because you're mutating one to probably persist (just) it again and the copy will be made at persistence time, but still a gotcha.)

> one under data and one under _containers

You're right, I should probably pop `_container` from the copied dict.

> Actually, where does _containers come from?

The mists of history; it's been present since the first revision in this repository. 

https://github.com/zopefoundation/persistent/blob/a6a18eaa8f347530646ce413f93f90a584ab2fb2/persistent/mapping.py#L124-L134

It must be something specific to `persistent`, though, because it doesn't show up in the [very first revision of `UserDict` from CPython](https://github.com/python/cpython/blame/bd40d7e69ffd53f36a315605fc849bb12ae1f55f/Lib/UserDict.py)

Reply:
It goes back to [this change from 2001](https://github.com/zopefoundation/ZODB/commit/c64281cf2830b569eed4f211630a8a61d22a0f0b#diff-b0f568e20f51129c10a096abad27c64a). Apparently "Standalone ZODB" used `data` but `ZODB` proper used `_container`. 

I wonder if there are any pickles that have been read-but-not-written (and are still being used read-only) with all the versions of ZODB and persistent since 2001. Surely not. Maybe this whole thing can be dropped.

Reply:
> Surely not

Hahahahaha.

I admire your optimism.

Reply:
Hope springs eternal. (I mean, there could be people entering university that are younger than those pickles!)

I added specific tests for the copy functionality when the ultra-legacy data layout is in place. I *hope* this PR is getting close to approval 😄